regression/cbmc-incr-smt2/bitvector-bitwise-operators/shift_right.c

```diff
@@ -0,0 +1,24 @@
+int main()
+{
+  // In C, whether a right-shift is arithmetic or logical depends on the
+  // original type being shifted. An unsigned value will be shifted to
+  // the right in a logical manner (this assigns `0` to the leftmost bit).
+  // If the type is signed, right shift will assign the sign bit to the
+  // leftmost digit.
+
+  int bitmask_2 = 0x2;
+  int bitmask_3 = 0x3;
+
+  int first;
+  unsigned int second;
+
+  int place;
+  __CPROVER_assume(place >= 1);
+
+  int result_signed = first >> place;
+  unsigned int result_unsigned = second >> place;
```

Comment:
We should test that `result_unsigned` is as expected in the case where the most significant bit of `second` is set. This is because this is the case where we would get the incorrect result if `sign_extend` is used instead of `zero_extend`.

Reply:
My apologies, I meant arithmetic vs logical shift right, not `xxx_extend`. I was mixed up with the casting conversions I had been working on. I think arithmetic shift right `bvashr` should be used for signed integers and logical right shift `bvlshr` should be used for unsigned integers, based on the signedness of the left hand operand. But I am not certain.
We should confirm the correct operation using a concrete execution of a compiled example program where a right shift is applied to a signed example and a unsigned example where the most significant bit is set and then the result is printed to console. We can then base the test on the observed behaviour. Examples where the most significant bit is set are interesting because the results are different based on the kind of shift carried out.

Reply:
The only way for me to verify right now which operation is being deployed is by increasing the verbosity level to something very high and try to match the log produced for the operation sent to the solver.

Is this a solution that you think is appropriate, or is there a better way to approach this?

Reply:
This should be solvable without resorting to turning up the verbosity. This could done by writing assertions which can be violated in the case of the expected shift, but not in the case of the incorrect shift.

Reply:
It could also be done based on turning on `--trace` and matching against the values in the trace.

Reply:
I don't want to depend on `trace` values too much, given that they are not exactly deterministic, and I have observed test failures because of alternative values in the trace before.

I'll see what I can do re assertions.

Reply:
If you write an assertion such that there is a single combination of inputs which cause the assertion to be violated then the values in the trace for that assertion are deterministic.

Reply:
lgtm